### ISSUE TITLE
Intermittent APIConnectionError(message='Could not verify Stripe's SSL certificate.  Please make sure that your network is not intercepting certificates

### ISSUE THREAD

**Issue by jayvdb** [stripe/stripe-python#660]
Over the last few days I have started seeing this error occurring, intermittently.  I am using the default http client which is choosing `requests`.  I set `stripe.verify_ssl_certs = False` but it is still happening.

```py
/usr/lib/python3.8/site-packages/urllib3/connectionpool.py:997: InsecureRequestWarning: Unverified HTTPS request is being made to host 'api.stripe.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  warnings.warn(
[07:11:33][ERROR] oscar_apps.checkout.facade facade.py:get_customer:108 | connected customer retrieve error: APIConnectionError(message='Could not verify Stripe\'s SSL certificate.  Please make sure that your\nnetwork is not intercepting certificates.  If this problem persists,\nlet us know at support@stripe.com.\n\n(Network error: SSLError: HTTPSConnectionPool(host=\'api.stripe.com\', port=443): Max retries exceeded with url: /v1/customers?email=***** (Caused by SSLError(SSLError("read error: Error([(\'SSL routines\', \'SSL_shutdown\', \'shutdown while in init\')])"))))', http_status=None, request_id=None)
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/urllib3/connectionpool.py", line 665, in urlopen
    httplib_response = self._make_request(
  File "/usr/lib/python3.8/site-packages/urllib3/connectionpool.py", line 421, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/usr/lib/python3.8/site-packages/urllib3/connectionpool.py", line 416, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib64/python3.8/http/client.py", line 1322, in getresponse
    response.begin()
  File "/usr/lib64/python3.8/http/client.py", line 303, in begin
    version, status, reason = self._read_status()
  File "/usr/lib64/python3.8/http/client.py", line 264, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib64/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3.8/site-packages/urllib3/contrib/pyopenssl.py", line 332, in recv_into
    raise ssl.SSLError("read error: %r" % e)
ssl.SSLError: ("read error: Error([('SSL routines', 'SSL_shutdown', 'shutdown while in init')])",)
```

**Comment by jayvdb** [comment]
This seems to make the problem disappear
```py
stripe.default_http_client = stripe.http_client.PycurlClient(
    verify_ssl_certs=False, proxy=False,
)
```
Note I am in Indonesia, not using a VPN atm, and they do funky stuff with network traffic, especially lots of HTTP sites, and usually only a few specific HTTPS sites (e.g. reddit).

**Comment by jleclanche** [comment]
@jayvdb that sounds like an issue with your ISP indeed; the code is most likely correct in crying out wolf.

But I do hope you're just playing with this and not setting `verify_ssl_certs=False` with production keys? Putting aside the security risks, you will not be [PCI compliant](https://stripe.com/guides/pci-compliance) if you do that (cf. "Encrypt transmission of cardholder data across open or public networks").

**Comment by jleclanche** [comment]
Looking at the code, it seems like `stripe.verify_ssl_certs = False` is not enough because the connection is being interrupted. That's no longer a certificate verification issue by that point but rather a connectivity error. (Which it's sounding a bit like is related to attempted eavesdropping)

**Comment by jayvdb** [comment]
This is just the development environment, connected to Stripe in test mode, pre-go-live.  It is odd that using pycurl instead of requests(urllib3) fixes the problem, but speaks in favour of pycurl which isnt a surprise,  I am happy enough with it as a workaround while it continues to work.  As an aside I am also using dj-stripe, so I needed to put my workaround very early in startup sequence - might be helpful if dj-stripe had an option to use pycurl instead of requests.  Leave that with you.  If you think it is a good idea, I could work on it later this week.

**Comment by jleclanche** [comment]
> It is odd that using pycurl instead of requests(urllib3) fixes the problem, but speaks in favour of pycurl which isnt a surprise

I would not be so sure :) It might be a bug in pycurl/libcurl that it doesn't give up. I'm sure both the libcurl and requests developers would be interested in an [HAR](https://en.wikipedia.org/wiki/HAR_\(file_format\)) of a request that breaks if it could be reproduced reliably or some such.

> might be helpful if dj-stripe had an option to use pycurl instead of requests

It's worth considering, but seems like a rare use case. How did you add your workaround? (We can take this discussion to a dj-stripe issue if you like)

**Comment by jayvdb** [comment]
I located the source.  requests in venvs was unexpected using normal DNS due to a minor system change which hadn't taken effect until a restart a few days ago which coincided with when I noticed it starting to occur; `pycurl` was detecting and using my local DNS over HTTPS server.  As a result I can't even be confident that this is a new Indonesian govt thing; I've noticed increasing messing around with DNS over the last year, hence patching most of my software to use DNS over HTTPS.  Still seems odd they would be including Stripe in their monitoring/whatever.

I added my pycurl workaround in settings.py, so it occurred before the Django app registry took over.  Agreed it is rare, and now even I don't *need* it, as I could get requests working with a bit more fiddling.  And we definitely have no need for it in production, which isnt located in Indonesia (very few sane ppl host in Indonesia).
